### DISCUSSION TITLE
Are structs always considered concrete?

### DISCUSSION THREAD

**Discussion by NReilingh** [cuelang/cue#890] [Q&A] [answered]
The [Language Specification](https://cuelang.org/docs/references/spec/) says

> A value is concrete if it is either an atom, or a struct all of whose field values are themselves concrete, recursively.

Yet I've come across a few things that seemingly contradict this:

#822 

> Things are a bit tricker when requiring fields of structs or lists to be concrete. As these are always concrete, all values would pass.

https://github.com/cuelang/cue/discussions/884#discussioncomment-578190

> It is working as expected because, [as documented](https://pkg.go.dev/cuelang.org/go@v0.0.6/cue#Value.IsConcrete), a struct is always considered concrete.
> 
> As per the doc:
> 
>> It does not verify that values of lists or structs are concrete themselves.

The spec seems to suggest that a struct is only concrete if all of its values are concrete, but the other materials say that structs (and lists?) are always concrete, regardless of the concreteness of the values that they contain.

Am I misunderstanding something, or is there a disconnect here?

**Comment by mpvl** [comment, marked as answer]
Good catch!

The spec is incorrect here. So it is a documentation bug.

**Reply by NReilingh** [reply]
Thanks! I'm curious about why that is -- the language spec definition seems like it would be the intuitive one to me. Somewhat related to this, perhaps, may also be how incompleteness relates to concreteness: I think again, intuitively, I would expect "completeness" and "concreteness" to be synonymous, and for incompleteness and concreteness to be mutually exclusive, but I suspect that may not be entirely accurate (for example, say, if a given struct is simultaneously concrete and incomplete).

Anyway I think it would be great if part of the docs gave some clarity to these details, since the terminology is used extensively throughout the project.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/890.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
